### PR TITLE
Python: update ToolCallBehavior and rename to FunctionCallBehavior to match dotnet and extended capabilities

### LINKED ISSUE #5414: Python: further implement parts of ToolCallBehavior to be in sync with dotnet
The dotnet implementation of ToolCallBehavior has additional components to it, which should be added to python.

### LINKED ISSUE #5447: Python: impose limit on num tools for tool calling
**Describe the bug**
OpenAI supports a max of 64 tools. We need to make sure we impose this limit in our settings.

### LINKED ISSUE #5727: Python: extend OpenAI ToolCallBehavior to match dotnet

### PR DESCRIPTION
**Pull request by eavanvalkenburg**
### Motivation and Context

Updated ToolCallBehavior to match the approach in dotnet
Closes #5727 
Closes #5447
Closes #5414

### Description

Extends ToolCallBehavior class with fields:

- enable_kernel_functions
- max_use_attempts
- allow_any_request_kernel_function

Added subclasses of TCB:
- KernelFunctions, the configure_options set tools to all functions in the kernel and tool_choice to auto
- EnabledFunctions, created with a filter dict, sets tool_choice to auto and tools to the filtered list
- RequiredFunction, created with a function fully qualified name (plugin-function), sets tool_choice to that name and adds the definition of just that tool to tools.

Methods:
- configure_options(kernel, exection_settings)
  - This sets the execution settings depending on the field in toolcallbehavior
  - Does nothing in the default ToolCallBehavior class, so there you have to manually set tools and tool_choice

ClassMethods:
- AutoInvokeKernelFunctions, returns KernelFunctions class with max_auto_invoke_attempts set to 5 (default)
- EnabelKernelFunctions, return KernelFunctions but with max_auto_invoke_attempts set to 0, disabling auto invoke, but it might return toolcalls from the model
- EnableFunctions, takes the filter and a auto_invoke params and returns a EnabledFunctions class, if auto_invoke == True then it will auto invoke, otherwise it wont.
- RequiredFunctions, returns RequiredFunction class with either max_invoke_attempts 0 or 1 depending on auto_invoke param

Changed OpenAIChatPromptExecutionSettings to have a field called tool_call_behavior instead of max_invoke_attempts and auto_invoke_tool_calls fields.
Some changes in openai_chat_completion_base to handle this.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/README.md">Py3.8 Test Coverage</a><details><summary>Python 3.8 Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>semantic_kernel</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py">kernel.py</a></td><td>261</td><td>28</td><td>89%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L122">122</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L124">124</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L182-L185">182&ndash;185</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L187-L191">187&ndash;191</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L193-L196">193&ndash;196</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L203">203</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L210-L219">210&ndash;219</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L254-L257">254&ndash;257</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L267-L272">267&ndash;272</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L277">277</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L281-L284">281&ndash;284</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/kernel.py#L466-L467">466&ndash;467</a></td></tr><tr><td colspan="5"><b>semantic_kernel/connectors/ai/open_ai</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py">utils.py</a></td><td>46</td><td>21</td><td>54%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L27-L28">27&ndash;28</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L51-L55">51&ndash;55</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L62-L63">62&ndash;63</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L104">104</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L142">142</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L144">144</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L146">146</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L148">148</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L150">150</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/utils.py#L155-L162">155&ndash;162</a></td></tr><tr><td colspan="5"><b>semantic_kernel/connectors/ai/open_ai/services</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py">open_ai_chat_completion_base.py</a></td><td>171</td><td>62</td><td>64%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L112">112</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L134-L137">134&ndash;137</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L139-L142">139&ndash;142</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L159-L162">159&ndash;162</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L172-L175">172&ndash;175</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L186-L209">186&ndash;209</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L218-L220">218&ndash;220</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L237-L239">237&ndash;239</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L262">262</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L270">270</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L276-L282">276&ndash;282</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L293-L299">293&ndash;299</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L335-L341">335&ndash;341</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L359">359</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L378-L380">378&ndash;380</a></td></tr><tr><td colspan="5"><b>semantic_kernel/functions</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/functions/kernel_function.py">kernel_function.py</a></td><td>73</td><td>2</td><td>97%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/functions/kernel_function.py#L142">142</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/functions/kernel_function.py#L161">161</a></td></tr><tr><td colspan="5"><b>semantic_kernel/planners/function_calling_stepwise_planner</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py">function_calling_stepwise_planner.py</a></td><td>102</td><td>38</td><td>63%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py#L122-L185">122&ndash;185</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py#L219-L224">219&ndash;224</a></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner_result.py">function_calling_stepwise_planner_result.py</a></td><td>15</td><td>2</td><td>87%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner_result.py#L6">6</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/9ece95b62d76cb2b45bac671e9ef8ba61c4276c2/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner_result.py#L28">28</a></td></tr><tr><td><b>TOTAL</b></td><td><b>5578</b></td><td><b>998</b></td><td><b>82%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python 3.8 Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 1211 | 18 :zzz: | 0 :x: | 0 :fire: | 21.318s :stopwatch: |

File: python/samples/kernel-syntax-examples/chat_gpt_api_function_calling.py [hunk 1]
```diff
@@ -163,16 +162,16 @@ async def chat() -> bool:
         print("\n\nExiting chat...")
         return False
 
-    stream = True
+    stream = False
```
Comment (juliomenendez) [inline python/samples/kernel-syntax-examples/chat_gpt_api_function_calling.py:165]: Is this change intended?

**Review by juliomenendez — APPROVED** [review]

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/README.md">Py3.10 Test Coverage</a><details><summary>Python 3.10 Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>semantic_kernel</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py">kernel.py</a></td><td>290</td><td>55</td><td>81%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L122">122</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L124">124</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L182-L185">182&ndash;185</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L187-L191">187&ndash;191</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L193-L196">193&ndash;196</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L203">203</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L210-L219">210&ndash;219</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L256-L259">256&ndash;259</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L269-L274">269&ndash;274</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L279">279</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L283-L286">283&ndash;286</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L468-L469">468&ndash;469</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/kernel.py#L744-L773">744&ndash;773</a></td></tr><tr><td colspan="5"><b>semantic_kernel/connectors/ai</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py">function_call_behavior.py</a></td><td>49</td><td>19</td><td>61%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py#L35-L39">35&ndash;39</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py#L54">54</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py#L67">67</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py#L77">77</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py#L89">89</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py#L100-L102">100&ndash;102</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py#L112-L114">112&ndash;114</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/function_call_behavior.py#L124-L130">124&ndash;130</a></td></tr><tr><td colspan="5"><b>semantic_kernel/connectors/ai/open_ai/contents</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/contents/function_call.py">function_call.py</a></td><td>38</td><td>4</td><td>89%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/contents/function_call.py#L41-L44">41&ndash;44</a></td></tr><tr><td colspan="5"><b>semantic_kernel/connectors/ai/open_ai/services</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py">open_ai_chat_completion_base.py</a></td><td>169</td><td>65</td><td>62%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L126">126</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L144">144</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L149-L152">149&ndash;152</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L154-L157">154&ndash;157</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L174-L177">174&ndash;177</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L187-L190">187&ndash;190</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L201-L224">201&ndash;224</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L233-L235">233&ndash;235</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L252-L254">252&ndash;254</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L277">277</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L285">285</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L291-L297">291&ndash;297</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L308-L314">308&ndash;314</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L333-L336">333&ndash;336</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L368">368</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L387-L389">387&ndash;389</a></td></tr><tr><td colspan="5"><b>semantic_kernel/functions</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/functions/kernel_function.py">kernel_function.py</a></td><td>76</td><td>3</td><td>96%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/functions/kernel_function.py#L143">143</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/functions/kernel_function.py#L162">162</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/functions/kernel_function.py#L257">257</a></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/functions/utils.py">utils.py</a></td><td>8</td><td>5</td><td>38%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/functions/utils.py#L16-L20">16&ndash;20</a></td></tr><tr><td colspan="5"><b>semantic_kernel/planners/function_calling_stepwise_planner</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py">function_calling_stepwise_planner.py</a></td><td>99</td><td>35</td><td>65%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py#L122-L186">122&ndash;186</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py#L220-L225">220&ndash;225</a></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner_result.py">function_calling_stepwise_planner_result.py</a></td><td>16</td><td>2</td><td>88%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner_result.py#L9">9</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/1011fce3cbdf7aee6c16c9b747f6ff38debd24ed/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner_result.py#L29">29</a></td></tr><tr><td><b>TOTAL</b></td><td><b>5756</b></td><td><b>1027</b></td><td><b>82%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python 3.10 Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 1211 | 1 :zzz: | 0 :x: | 0 :fire: | 35.610s :stopwatch: |

**Review by juliomenendez — APPROVED** [review]

File: python/semantic_kernel/connectors/ai/open_ai/services/utils.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+
+def update_settings_from_function_call_configuration(
+    function_call_configuration: "FunctionCallConfiguration", settings: "OpenAIChatPromptExecutionSettings"
+) -> None:
+    """Update the settings from a FunctionCallConfiguration."""
+    if function_call_configuration.required_functions:
+        if len(function_call_configuration.required_functions) > 1:
+            logger.warning("Multiple required functions are not supported. Using the first required function.")
```
Comment (alliscode) [inline python/semantic_kernel/connectors/ai/open_ai/services/utils.py:32]: What is the scenario where this would happen? Is it if there are multiple functions that have the same name as `function_fully_qualified_name`.

**Review by alliscode — APPROVED** [review]

File: python/semantic_kernel/connectors/ai/open_ai/services/utils.py (hunk 2 above)
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/open_ai/services/utils.py:32]: This looks like it's a warning if the LLM returns back multiple functions per function call/tool call. Right now, only expect there to be one function call (ID) mapping to one function call. In the event it ever returned back more than one, we'd get this warning.
